examples/generative/finetune_stable_diffusion.py

```diff
@@ -0,0 +1,507 @@
+            # Know more about the magic number here:
+            # https://keras.io/examples/generative/fine_tune_via_textual_inversion/
+            latents = latents * 0.18215
+
+            # Sample noise that we'll add to the latents.
+            noise = tf.random.normal(tf.shape(latents))
+
+            # Sample a random timestep for each image.
+            timesteps = tnp.random.randint(
+                0, self.noise_scheduler.train_timesteps, (batch_size,)
+            )
+
+            # Add noise to the latents according to the noise magnitude at each timestep
+            # (this is the forward diffusion process).
+            noisy_latents = self.noise_scheduler.add_noise(
+                tf.cast(latents, noise.dtype), noise, timesteps
+            )
+
+            # Get the target for loss depending on the prediction type
+            # just the sampled noise for now.
+            target = noise  # noise_schedule.predict_epsilon == True
+
+            # Predict the noise residual and compute loss.
+            timestep_embedding = tf.map_fn(
+                lambda t: self.get_timestep_embedding(t), timesteps, dtype=tf.float32
+            )
+            timestep_embedding = tf.squeeze(timestep_embedding, 1)
+            model_pred = self.diffusion_model(
+                [noisy_latents, timestep_embedding, encoded_text], training=True
+            )
+            loss = self.compiled_loss(target, model_pred)
+            if self.use_mixed_precision:
+                loss = self.optimizer.get_scaled_loss(loss)
+
+        # Update parameters of the diffusion model.
+        trainable_vars = self.diffusion_model.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+        if self.use_mixed_precision:
+            gradients = self.optimizer.get_unscaled_gradients(gradients)
+        gradients = [tf.clip_by_norm(g, self.max_grad_norm) for g in gradients]
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        return {m.name: m.result() for m in self.metrics}
+
+    def get_timestep_embedding(self, timestep, dim=320, max_period=10000):
+        half = dim // 2
+        log_max_preiod = tf.math.log(tf.cast(max_period, tf.float32))
+        freqs = tf.math.exp(
+            -log_max_preiod * tf.range(0, half, dtype=tf.float32) / half
+        )
+        args = tf.convert_to_tensor([timestep], dtype=tf.float32) * freqs
+        embedding = tf.concat([tf.math.cos(args), tf.math.sin(args)], 0)
+        embedding = tf.reshape(embedding, [1, -1])
+        return embedding
+
+    def sample_from_encoder_outputs(self, outputs):
+        mean, logvar = tf.split(outputs, 2, axis=-1)
+        logvar = tf.clip_by_value(logvar, -30.0, 20.0)
+        std = tf.exp(0.5 * logvar)
+        sample = tf.random.normal(tf.shape(mean), dtype=mean.dtype)
+        return mean + std * sample
+
+    def save_weights(self, filepath, overwrite=True, save_format=None, options=None):
```

Comment:
ultimately do you still want to save the encoder part, so it could load and run inference later?

Reply:
See the inference code later to see how that might not be required. 

We just do:

```py
pokemon_model = keras_cv.models.StableDiffusion(
    img_width=img_width, img_height=img_height
)
# We just reload the weights of the fine-tuned diffusion model.
pokemon_model.diffusion_model.load_weights(weights_path)
```

Reply:
Makes sense